File: src/main/java/com/mkreidl/ephemeris/sky/PlanetRiseSetCalculator.java
package com.mkreidl.ephemeris.sky;

import com.mkreidl.ephemeris.Distance;
import com.mkreidl.ephemeris.Position;
import com.mkreidl.ephemeris.Time;
import com.mkreidl.ephemeris.solarsystem.Body;
import com.mkreidl.ephemeris.solarsystem.SolarSystem;

public class PlanetRiseSetCalculator extends RiseSetCalculator
{
    private static final int MAX_ITERATION = 5;

    private final SolarSystem solarSystem;
    private final Body body;
    private final Position position = new Position();

    private long timeMillisPrevious;

    private Boolean wasVisibleBefore;
    private Boolean isVisibleNow;
    private boolean isCrossing;
    private long precisionMs = 1000;

    public static PlanetRiseSetCalculator of( SolarSystem solarSystem, Body body )
    {
        return new PlanetRiseSetCalculator( solarSystem, body );
    }

    private PlanetRiseSetCalculator( SolarSystem solarSystem, Body body )
    {
        this.solarSystem = solarSystem;
        this.body = body;
    }
    
    public void setPrecision( long precisionMs )
    {
      this.precisionMs = precisionMs;
    }

    @Override
    public boolean compute( long timeMillisStart )
    {
        super.setStartTime( timeMillisStart );
        isVisibleNow = null;
        for ( int n = 0; n < MAX_ITERATION; ++n )
        {
            timeMillisPrevious = time.getTime();
            computeTopocentricPosition();
            if ( !isCrossing || !adjustTime() )
                if ( mode == EventType.RISE && hasAppeared() || mode == EventType.SET && hasVanished() )
                    time.setTime( ( time.getTime() + timeMillisPrevious ) / 2 );
                else
                    return compute( searchOrbitCrossingHorizon() );
            if ( Math.abs( time.getTime() - timeMillisPrevious ) < precisionMs )
                return true;
        }
        return false;
    }

    private long searchOrbitCrossingHorizon()
    {
        long searchIncrement = 30 * Time.MILLIS_PER_SIDEREAL_DAY * ( lookupDirection == LookupDirection.FORWARD ? 1 : -1 );
        while ( !isCrossing )
        {
            timeMillisPrevious = time.getTime();
            time.addMillis( searchIncrement );
            computeTopocentricPosition();
        }
        while ( Math.abs( searchIncrement ) > Time.MILLIS_PER_HOUR )
        {
            if ( !isCrossing )
                timeMillisPrevious = time.getTime();
            searchIncrement /= 2;
            time.setTime( timeMillisPrevious + searchIncrement );
            computeTopocentricPosition();
        }
        return time.getTime();
    }

    private void computeTopocentricPosition()
    {
        wasVisibleBefore = isVisibleNow;
        solarSystem.compute( time, Body.EARTH );
        solarSystem.compute( time, body );
        solarSystem.getEphemerides( body, position );
        position.setTimeLocation( time, geographicLocation );
        position.get( topocentric, Position.CoordinatesCenter.TOPOCENTRIC );
        final double apparentRadius = body.getRADIUS_MEAN_M() / topocentric.distance( Distance.m );
        virtualHorizonDeg = OPTICAL_HORIZON_DEG - Math.toDegrees( apparentRadius );
        isVisibleNow = topocentric.lat >= virtualHorizonDeg;
        isCrossing = isCrossing();
    }

    private boolean hasAppeared()
    {
        return wasVisibleBefore != null && isVisibleNow && !wasVisibleBefore;
    }

    private boolean hasVanished()
    {
        return wasVisibleBefore != null && !isVisibleNow && wasVisibleBefore;
    }
}

File: src/main/java/com/mkreidl/ephemeris/sky/PlanetRiseSetCalculator.kt
package com.mkreidl.ephemeris.sky

import com.mkreidl.ephemeris.MILLIS_PER_HOUR
import com.mkreidl.ephemeris.MILLIS_PER_SIDEREAL_DAY
import com.mkreidl.ephemeris.solarsystem.Body
import com.mkreidl.ephemeris.solarsystem.SolarSystem
import com.mkreidl.ephemeris.time.Instant

class PlanetRiseSetCalculator(
        private val solarSystem: SolarSystem,
        private val body: Body,
        mode: EventType,
        lookupDirection: LookupDirection
) : RiseSetCalculator(mode, lookupDirection) {

    private var timeMillisPrevious = 0L
    private var wasVisibleBefore: Boolean? = null
    private var isVisibleNow = false
    private var crossing = false
    private var precisionMs = 1000L

    override fun compute(): Boolean {
        for (n in 0 until MAX_ITERATION) {
            timeMillisPrevious = time.epochMilli
            computeTopocentricPosition()
            if (!crossing || !adjustTime())
                if (mode == RiseSetCalculator.EventType.RISE && hasAppeared() || mode == RiseSetCalculator.EventType.SET && hasVanished()) {
                    time = Instant.ofEpochMilli((time.epochMilli + timeMillisPrevious) / 2)
                } else {
                    searchOrbitCrossingHorizon()
                    return compute()
                }
            if (Math.abs(time.epochMilli - timeMillisPrevious) < precisionMs) {
                return true
            }
            wasVisibleBefore = isVisibleNow
        }
        return false
    }

    private fun searchOrbitCrossingHorizon() {
        var searchIncrement = (30 * MILLIS_PER_SIDEREAL_DAY * lookupDirection.sign).toLong()
        while (!crossing) {
            timeMillisPrevious = time.epochMilli
            startTime = time.addMillis(searchIncrement)
            computeTopocentricPosition()
        }
        while (Math.abs(searchIncrement) > MILLIS_PER_HOUR) {
            if (!crossing)
                timeMillisPrevious = startTime.epochMilli
            searchIncrement /= 2
            startTime = Instant.ofEpochMilli(timeMillisPrevious + searchIncrement)
            computeTopocentricPosition()
        }
    }

    private fun computeTopocentricPosition() {
        solarSystem.computeSingle(time, body = body)
        val geocentric = solarSystem.getTrueEquatorialGeocentric(body).position
        topocentric = topos.computeTopocentricFromTrueEquatorial(geocentric).spherical
        virtualHorizon = RiseSetCalculator.OPTICAL_HORIZON - body.RADIUS_MEAN_M / topocentric.dst
        isVisibleNow = topocentric.lat >= virtualHorizon
        crossing = eventHappensToday()
    }

    private fun hasAppeared() = wasVisibleBefore == false && isVisibleNow
    private fun hasVanished() = wasVisibleBefore == true && !isVisibleNow

    companion object {
        private const val MAX_ITERATION = 5
    }
}